pkg/api/v1/summary.go

```diff
@@ -0,0 +1,132 @@
+	summarypb "github.com/GoogleCloudPlatform/testgrid/pb/summary"
+	"github.com/GoogleCloudPlatform/testgrid/pkg/summarizer"
+)
+
+var (
+	// These should be in line with the TabStatus enum in state.proto.
+	tabStatusStr = map[summarypb.DashboardTabSummary_TabStatus]string{
+		summarypb.DashboardTabSummary_PASS:       "PASSING",
+		summarypb.DashboardTabSummary_FAIL:       "FAILING",
+		summarypb.DashboardTabSummary_FLAKY:      "FLAKY",
+		summarypb.DashboardTabSummary_STALE:      "STALE",
+		summarypb.DashboardTabSummary_BROKEN:     "BROKEN",
+		summarypb.DashboardTabSummary_PENDING:    "PENDING",
+		summarypb.DashboardTabSummary_ACCEPTABLE: "ACCEPTABLE",
+	}
+)
+
+// fetchSummary returns the summary struct as defined in summary.proto.
+// Returns an error iff the scope refers to non-existent bucket OR server fails to read the summary.
+func (s *Server) fetchSummary(ctx context.Context, scope, dashboard string) (*summarypb.DashboardSummary, error) {
+	configPath, _, err := s.configPath(scope)
+
+	summaryPath, err := summarizer.SummaryPath(*configPath, s.SummaryPathPrefix, dashboard)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create the summary path: %v", err)
+	}
+
+	summary, _, _, err := summarizer.ReadSummary(ctx, s.Client, *summaryPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to download summary at %v: %v", summaryPath.String(), err)
+	}
+
+	return summary, nil
+}
+
+// ListTabSummaries returns the list of tab summaries for the particular dashboard.
+// Returns an error iff dashboard does not exist OR the server can't read summaries from GCS bucket.
+func (s *Server) ListTabSummaries(ctx context.Context, req *apipb.ListTabSummariesRequest) (*apipb.ListTabSummariesResponse, error) {
+	ctx, cancel := context.WithTimeout(ctx, s.Timeout)
+	defer cancel()
+
+	scope := req.GetScope()
+	cfg, err := s.getConfig(ctx, logrus.WithContext(ctx), scope)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch config from {%q}: %v", scope, err)
+	}
+
+	cfg.Mutex.RLock()
+	defer cfg.Mutex.RUnlock()
+
+	dashboardKey := config.Normalize(req.GetDashboard())
+	if _, ok := cfg.NormalDashboard[dashboardKey]; !ok {
+		return nil, fmt.Errorf("dashboard {%q} not found", dashboardKey)
+	}
+
+	summary, err := s.fetchSummary(ctx, scope, dashboardKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fetch summary for dashboard {%q}: %v", dashboardKey, err)
+	}
+
+	if summary == nil {
+		return nil, fmt.Errorf("summary for dashboard {%q} not found.", dashboardKey)
+	}
+
+	var resp apipb.ListTabSummariesResponse
+	for _, tabSummary := range summary.TabSummaries {
+		ts := &apipb.TabSummary{
+			DashboardName:         tabSummary.DashboardName,
+			TabName:               tabSummary.DashboardTabName,
+			OverallStatus:         tabStatusStr[tabSummary.OverallStatus],
+			DetailedStatusMessage: tabSummary.Status,
+			LastRunTimestamp: &timestamp.Timestamp{
+				Seconds: int64(tabSummary.LastRunTimestamp),
```

Comment:
nit: It looks like tabSummary.LastRunTimestamp is a float. This intrinsically rounds the timestamp to the nearest second, which is probably fine: don't feel like you need to do this in this PR.

More precisely, though, you may want to convert the fractional part of the float into nanos. Example 5 of this kind of illustrates what I mean: https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Timestamp

Reply:
Sounds good, I will add this as a TODO. Thanks!